provider/internal/builtin/service_client.go

```diff
@@ -284,3 +262,85 @@ func findFilesMatchingPattern(root, pattern string) ([]string, error) {
 	})
 	return matches, err
 }
+
+type walkResult struct {
+	positionParams protocol.TextDocumentPositionParams
+	match          string
+}
+
+func parallelWalk(location string, regex *regexp.Regexp) ([]walkResult, error) {
+	var positions []walkResult
+	positionsChan := make(chan walkResult)
+	wg := &sync.WaitGroup{}
+
+	go func() {
+		err := filepath.Walk(location, func(path string, f os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if f.Mode().IsRegular() {
+				wg.Add(1)
+				go processFile(path, regex, positionsChan, wg)
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return
+		}
+
+		wg.Wait()
+		close(positionsChan)
+	}()
+
+	for pos := range positionsChan {
+		positions = append(positions, pos)
+	}
+
+	return positions, nil
+}
+
+func processFile(path string, regex *regexp.Regexp, positionsChan chan<- walkResult, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	content, err := os.ReadFile(path)
```

Comment:
Little concerned about memory here. I think we don't have limits on workers either. Should we have another check to switch `forceFullFileScan` if file size is beyond a certain value?

Reply:
is the idea that going line by line and bringing it into the buffer will reduce the memory. 

The problem could arise if we have a jar file that is GB big, but even then, I don't think that going line by line is going to reduce the size of memory usage because do those compacted files have line breaks?

Reply:
@shawn-hurley I am suggesting that, beyond a certain size, we shouldn't `ReadFile` and only do the Scan(). I wasn't really worried about binaries, but in general, big files (for instance, a db seed). 

If the line break isn't present within first 64K, Scanner() will fail I think. So that's an edge case to that I didn't think about earlier.

Reply:
If there's a way to get rid of `ReadFile` and only do scan in chunks of pre-determined size would be even better. That will affect the accuracy unless we have a window mechanism.

Reply:
I think I am suggesting that we live with this for now; if we see an issue in testing all the stuff, then let's come back. I can add a TODO that this may need to be changed.

Thoughts?

Reply:
I am fine with addressing this as a TODO. Thank you @shawn-hurley this looks great

Reply:
Captured concern in Issue: https://github.com/konveyor/analyzer-lsp/issues/433